sw/device/tests/sim_dv/ast_clk_rst_inputs.c

```diff
@@ -0,0 +1,577 @@
+  }
+}
+
+void ast_enter_sleep_states_and_check_functionality(
+    dif_pwrmgr_domain_config_t pwrmgr_config,
+    dif_entropy_src_config_t entropy_src_config, uint32_t alert_idx) {
+  bool deepsleep;
+  int read_fifo_depth_val;
+  bool clock_check_status;
+  uint32_t unhealthy_fifos, errors, alerts;
+
+  const dif_edn_t edn0 = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_EDN0_BASE_ADDR)};
+  const dif_edn_t edn1 = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_EDN1_BASE_ADDR)};
+
+  if ((pwrmgr_config & (~kDifPwrmgrDomainOptionUsbClockInActivePower)) == 0) {
+    deepsleep = true;
+  } else {
+    deepsleep = false;
+  }
+
+  if (pwrmgr_testutils_is_wakeup_reason(&pwrmgr, 0)) {
+    CHECK_DIF_OK(dif_entropy_src_set_enabled(&entropy_src, kDifToggleDisabled));
+
+    const dif_entropy_src_fw_override_config_t fw_override_config = {
+        .entropy_insert_enable = true,
+        .buffer_threshold = kEntropyFifoBufferSize,
+    };
+
+    CHECK_DIF_OK(dif_entropy_src_fw_override_configure(
+        &entropy_src, fw_override_config, kDifToggleEnabled));
+
+    CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                           kDifToggleEnabled));
+
+    // Verify that the FIFO depth is non-zero via SW - indicating the reception
+    // of data over the AST RNG interface.
+    IBEX_SPIN_FOR(read_fifo_depth(&entropy_src) > 0, 1000);
+
+    // test recoverable event
+    test_event(alert_idx, kDifToggleDisabled, kAlertSet);
+
+    // Enable all the AON interrupts used in this test.
+    rv_plic_testutils_irq_range_enable(
+        &rv_plic, kTopEarlgreyPlicTargetIbex0,
+        kTopEarlgreyPlicIrqIdAdcCtrlAonMatchDone,
+        kTopEarlgreyPlicIrqIdAdcCtrlAonMatchDone);
+    CHECK_DIF_OK(dif_pwrmgr_irq_set_enabled(&pwrmgr, 0, kDifToggleEnabled));
+
+    // Enter low power mode.
+    LOG_INFO("Issued WFI to enter sleep.");
+    // Setup low power.
+    rstmgr_testutils_pre_reset(&rstmgr);
+
+    if (!deepsleep) {
+      // read fifo depth before enter sleep mode
+      read_fifo_depth_val = read_fifo_depth(&entropy_src);
+    }
+
+    // Configure ADC
+    adc_setup(first_adc_setup);
+
+    // set sleep mode
+    pwrmgr_testutils_enable_low_power(&pwrmgr, kDifPwrmgrWakeupRequestSourceTwo,
+                                      pwrmgr_config);
+
+    wait_for_interrupt();
+
+    // Interrupt should have been serviced.
+    CHECK(interrupt_serviced);
+
+    interrupt_serviced = false;
+
+  } else if (pwrmgr_testutils_is_wakeup_reason(
+                 &pwrmgr, kDifPwrmgrWakeupRequestSourceTwo)) {
+    if (deepsleep) {
+      if (read_fifo_depth(&entropy_src) != 0)
+        LOG_ERROR("read_fifo_depth after reset=%0d should be 0",
+                  read_fifo_depth(&entropy_src));
+
+      first_adc_setup = false;
+      // Configure ADC after deep sleep
+      adc_setup(first_adc_setup);
+    }
+
+    CHECK_DIF_OK(dif_entropy_src_set_enabled(&entropy_src, kDifToggleDisabled));
```

Comment:
Some of this code seems the same as code near line 310. Move it to a function or run it before branching?